### DISCUSSION TITLE
Implementing companion specifications?

### DISCUSSION THREAD

**Discussion by honkomonk** [FreeOpcUa/opcua-asyncio#650] [Q&A]
Hi all,

I'm planing to implement a companion specification and wonder what would be the right way to do it using asyncua? Specifically I want to implement the MachineVision specification (`schemas/UA-Nodeset-master/MachineVision/Opc.Ua.MachineVision.NodeSet2.xml`). 

At the moment I see 3 ways to approach this:
1.) Hardcode it by hand. Works, but it's tedious, error-prone and more work when the specification gets updated.
2.) Import the XML specs and add the logic in Python. Would be my path of choice, but xml file appears to contain datastructures the xml-importer can't handle. If "this is the way" I might be able to have a deeper look to debbug the cause.
3.) Autogenerate the code. Maybe it's possible to use the `generate_*` scripts in the `/schemas/` directory to generate the code from the .xml definition?

Regards

**Comment by AndreasHeine** [comment]
typically you need to import the spec. xml and then you can instantiate the obejcttypes according to the spec.

https://github.com/umati/Sample-Server-asyncio

**Comment by AndreasHeine** [comment]
what errors did you got?

**Comment by oroulet** [comment]
Yes importing the XML is the way to go. And if the importer has bugs then you would help everyone by fixing them.
 XML importer code is not that good but if you have ideas to improve, you are welcome

**Comment by swamper123** [comment]
I had taken a look on [the specs xml](https://github.com/OPCFoundation/UA-Nodeset/blob/v1.04/MachineVision/Opc.Ua.MachineVision.NodeSet2.xml) and it hasn't got any other xml requirements to be imported first (only the basic itself, but this is done anyway).

Like Andreas said, you can take a look on the umati example or [this one from us](https://github.com/FreeOpcUa/opcua-asyncio/blob/3b974f1407fa4c90d14a13ae8b18e20e22aef8e1/examples/server-robotics.py) (just import the stuff after `init()` and before `start()`).

Since I touched the importer the last time, it is of course perfect and no errors ever happened(\*turn sarcasm off\*). ;^)

> but xml file appears to contain datastructures the xml-importer can't handle.

Please provide the log of the error you have. Either here or even better as an separate issue.

Also keep in mind, that UAMethods have to be linked with a Python function, before usage (otherwise they are empty/faulty).

**Comment by honkomonk** [comment]
Thanks for the quick replies!   The _umati_ server implementation looks interesting and might be a good starting point. I'll have a look.

@swamper123 Of course I assumed impeccable  functioning of the importer. That's why I asked what I might be doing wrong.  ;-)

For testing I used the `server-robotics.py` script and just dropped in the path to the other nodeset file. I'll open up a new Issue and post the code and stacktrace. It doesn't look trivial to me, but might be for someone knowing the inner workings of the importer.

The Issue: #651
